e2e/rgraph_update_action_test.go

```diff
@@ -0,0 +1,262 @@
+	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud/meta"
+	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud/rgraph"
+	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud/rgraph/exec"
+	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud/rgraph/rnode"
+	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud/rgraph/rnode/backendservice"
+	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud/rgraph/rnode/healthcheck"
+	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud/rgraph/workflow/plan"
+	"google.golang.org/api/compute/v1"
+)
+
+func buildBackendService(graphBuilder *rgraph.Builder, name string, hcID *cloud.ResourceID, port int64) (*cloud.ResourceID, error) {
+	bsID := backendservice.ID(testFlags.project, meta.GlobalKey(resourceName(name)))
+
+	bsMutResource := backendservice.NewMutableBackendService(testFlags.project, bsID.Key)
+	bsMutResource.Access(func(x *compute.BackendService) {
+		x.LoadBalancingScheme = "INTERNAL_SELF_MANAGED"
+		x.Protocol = "TCP"
+		x.Port = port
+		x.HealthChecks = []string{hcID.SelfLink(meta.VersionGA)}
+	})
+	bsResource, err := bsMutResource.Freeze()
+	if err != nil {
+		return nil, err
+	}
+
+	bsBuilder := backendservice.NewBuilder(bsID)
+	bsBuilder.SetOwnership(rnode.OwnershipManaged)
+	bsBuilder.SetState(rnode.NodeExists)
+	bsBuilder.SetResource(bsResource)
+
+	graphBuilder.Add(bsBuilder)
+	return bsID, nil
+}
+
+func buildHealthCheck(graphBuilder *rgraph.Builder, name string, checkIntervalSec int64) (*cloud.ResourceID, error) {
+	hcID := healthcheck.ID(testFlags.project, meta.GlobalKey(resourceName(name)))
+	hcMutRes := healthcheck.NewMutableHealthCheck(testFlags.project, hcID.Key)
+	hcMutRes.Access(func(x *compute.HealthCheck) {
+		x.CheckIntervalSec = checkIntervalSec
+		x.HealthyThreshold = 5
+		x.TimeoutSec = 6
+		x.Type = "HTTP"
+		x.HttpHealthCheck = &compute.HTTPHealthCheck{
+			RequestPath: "/",
+			Port:        int64(9376),
+		}
+	})
+	hcRes, err := hcMutRes.Freeze()
+	if err != nil {
+		return nil, err
+	}
+	hcBuilder := healthcheck.NewBuilder(hcID)
+	hcBuilder.SetOwnership(rnode.OwnershipManaged)
+	hcBuilder.SetState(rnode.NodeExists)
+	hcBuilder.SetResource(hcRes)
+	graphBuilder.Add(hcBuilder)
+	return hcID, nil
+}
+
+func TestHcUpdateWithBackendService(t *testing.T) {
+	ctx := context.Background()
+	graphBuilder := rgraph.NewBuilder()
+
+	hcID, err := buildHealthCheck(graphBuilder, "hc-test", 15)
+	if err != nil {
+		t.Fatalf("buildHealthCheck(_, hc-test, _) = %v, want nil", err)
+	}
+
+	bsID, err := buildBackendService(graphBuilder, "bs-e2e", hcID, 80)
+	if err != nil {
+		t.Fatalf("buildBackendService(_, bs-e2e, _, 80) = %v, want nil", err)
+	}
+
+	graph, err := graphBuilder.Build()
+	if err != nil {
+		t.Fatalf("graphBuilder.Build() = %v, want nil", err)
+	}
+	result, err := plan.Do(ctx, theCloud, graph)
+	if err != nil {
+		t.Fatalf("plan.Do(_, _, _) = %v, want nil", err)
+	}
+
+	var viz exec.GraphvizTracer
```

Comment:
nit: this is unused (I don't see the outputs saved anywhere). Do you plan to use it for debugging?

Reply:
done